dbt/adapters/snowflake/impl.py

```diff
@@ -147,23 +147,40 @@ def list_relations_without_caching(
         quote_policy = {"database": True, "schema": True, "identifier": True}
 
         columns = ["database_name", "schema_name", "name", "kind"]
-        for _database, _schema, _identifier, _type in results.select(columns):
-            try:
-                _type = self.Relation.get_relation_type(_type.lower())
-            except ValueError:
-                _type = self.Relation.External
+        if "is_dynamic" in results.column_names:
+            columns.append("is_dynamic")
+
+        for result in results.select(columns):
+            database, schema, identifier, relation_type = self._parse_list_relations_result(result)
             relations.append(
                 self.Relation.create(
-                    database=_database,
-                    schema=_schema,
-                    identifier=_identifier,
+                    database=database,
+                    schema=schema,
+                    identifier=identifier,
+                    type=relation_type,
                     quote_policy=quote_policy,
-                    type=_type,
                 )
             )
 
         return relations
 
+    def _parse_list_relations_result(self, result: agate.Row) -> Tuple[str, str, str, str]:
+        try:
+            database, schema, identifier, relation_type, is_dynamic = result
+        except ValueError:
+            database, schema, identifier, relation_type = result
+            is_dynamic = "N"
+
+        try:
+            relation_type = self.Relation.get_relation_type(relation_type.lower())
+        except ValueError:
+            relation_type = self.Relation.External
+
+        if relation_type == self.Relation.Table and is_dynamic == "Y":
```

Comment:
Why can't this just be placed on 169? Why it's own if?

My thinking with what I know: `is_dynamic` will be `Y` if it's not `N`, right? Or if there is something third value, do we handle that edge case or let it exist silently?

Reply:
There are a couple of reasons. I try to minimize what is inside of a `try`/`except` so that we're only testing the expected failure. I wouldn't want to bury an if in there that's dependent on the outcome of the thing that may fail. Also, we have not gotten a `RelationType` yet, we only have whatever string Snowflake put in their metadata table (which already changed once, and is part of why we're doing this change). This is the first place where we have cleared both of those issues.

The other alternative that I considered was to have two completely different parsing methods based on whether `is_dynamic` is available (basically if `2024_03` is active). I couldn't see a way to do that without creating a bunch of methods, nesting for loops, or adding non-intuitive arguments to methods. Additionally, this is only an issue until `2024_03` is mandatory, after which we can remove `try` block and replace it with the unpacking on line 169 since `is_dynamic` would be reliable at that point. The rest of this method stays the same.